### PR TITLE
[ContactStructuralMechanicsApplication] Adding compatibility in contact between quadrilaterals and triangles

### PR DESCRIPTION
**Pull request by loumalouomega**
This is a huge PR, but basically just affects ContactStructuralMechanicsApplication. This includes all the changes needed in order to make compatible contact between triangles and quadrilaterals. Includes tests for contact and mesh tying

### CODE REVIEW

**Comment by philbucher** [comment]
@loumalouomega I tried to review this but it is just too large:
- First of all git cannot even show the diff for a file!
~~~
90,515 additions, 6,075 deletions not shown because the diff is too large. Please use a local Git client to view these changes.
~~~

- Second PLEASE do cosmetic changes in a separate PR 
screenshot from 2018-09-11 09-20-51
All those changes are just cosmetic but one still has to look at all of them bcs some of them might not only be cosmetic...
This PR changes >300.000 lines changed so this would take me all day :D

Can you first do a PR with just the cosmetic changes and then one with the actual changes?

**Comment by loumalouomega** [comment]
The formulation is generated using AD, the resulting files are huge. The only ones to review are the scripts to generate it

**Comment by loumalouomega** [comment]
I can separate into smaller pieces, let's see

**Comment by loumalouomega** [comment]
I created #2845

**Comment by loumalouomega** [comment]
OK, after merging #2856 this PR just adds contact conditions

**Comment by loumalouomega** [comment]
Ping

**Comment by loumalouomega** [comment]
Ping

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h [hunk 1]
```diff
@@ -262,6 +262,69 @@ class DerivativesUtilities
...
 
+    /**
+     * @brief This method is used to compute the local increment of the normal
+     * @param rJacobian The jacobian on the GP
+     * @param rDNDe The local gradient
+     * @return The matrix containing the delta normals
+     * @note Not the mean, look in the contact utilities
+     * @note Hardcopied for performance
+     */
+    static inline array_1d<array_1d<double, 3>, TDim * TNumNodesMaster> GPDeltaNormalMaster(
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:273]: you sure you want to inline such a large function?

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h [hunk 2]
```diff
@@ -1339,6 +1403,44 @@ class DerivativesUtilities
...
 
+    /**
+     * @brief This method computes the increment of local coordinates
+     * @param rResult The solution obtained
+     * @param rDeltaPoint The increment of position in the points
+     * @param rThisGeometry The geometry considered (master)
+     * @param rThisNormal The normal of the geometry
+     * @note Hardcopied for performance
+     */
+    static inline void DeltaPointLocalCoordinatesMaster(
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:1414]: idem

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h [hunk 3]
```diff
@@ -1339,6 +1403,44 @@ class DerivativesUtilities
...
+            X(2, i) = rThisGeometry[i].Z();
+        }
+
+        const BoundedMatrix<double, 3, 2> DN = prod(X, rDNDe);
+
+        const BoundedMatrix<double, 2, 2> J = prod(trans(DN),DN);
+        double det_j = MathUtils<double>::DetMat<BoundedMatrix<double, 2, 2>>(J);
+        const BoundedMatrix<double, 2, 2> invJ = (std::abs(det_j) < ZeroTolerance) ? ZeroMatrix(2,2) : MathUtils<double>::InvertMatrix<2>(J, det_j);
+
+    #ifdef KRATOS_DEBUG
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:1435]: you sure abt this?

**Review by philbucher — APPROVED** [review]
I cannot review the cpp-files of the contact conditions bcs the diff is too large

I guess that you generate those using some sort of script anyway right?

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h (hunk 1 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:273]: It is not used in many places, but I can remove the inline in a future PR

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h (hunk 2 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:1414]: The same, for a future PR.
Which is the magnitude order of the inline more or less?

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h (hunk 3 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:1435]: About doing a check just in debug, yes

**Pull request merged**